### PR TITLE
wallet-core: Add functionality to create transaction

### LINKED ISSUE #2128: wallet-core: create transactions
The `wallet-core` should provide a function capable of creating transactions for both Phoenix and Moonlight, with support for delegating proof generation (i.e., "unproven transactions") to an external (remote) prover. This delegation should ensure that the transactions remain non-malleable by the prover, as per the current implementation.

Key requirements for this function include:

1. It should be versatile enough to handle Dusk token transfers, staking operations, and all functionalities currently supported by the genesis contracts on our network.
2. The function should accept generic call data, allowing for flexible transaction creation. For example, it should enable users to create a transaction for invoking the "foo" method on a specific contract address, passing a boolean argument and ensuring that the associated gas costs are covered.

### PR DESCRIPTION
**Pull request by moCello**
To be able to create transactions the rusk-prover needed to be updated as well.
The `rusk-prover::Prover` trait is replaced by the `execution_core::transfer::phoenix::Prove` trait.
With this change, the `rusk-prover` library now only holds the `LocalProver` implementation of the `Prove` trait and the feature structure was drastically simplified.

Additionally I took the liberty to also adjust the names of the cached `TxCircuit` circuit-profile, prover and verifier in order for them to better match their actual circuit`s name.

The rusk http prover request also changed from `"prove_execute"` to `"prove_tx_circuit"`.

Resolves #2128

### CODE REVIEW

File: wallet-core/src/transaction.rs
Comment (Daksh14) [inline wallet-core/src/transaction.rs:49]: a type alias here would be nice, its typed out multiple times too

File: wallet-core/src/transaction.rs
Comment (Daksh14) [inline wallet-core/src/transaction.rs:34]: The ProverClient trait would be nice to have but will be problematic for the FFI unless we do wasm imports. A function to replace this would be nice and all the other transaction functions return an unproven transaction

**Review by Daksh14 — CHANGES_REQUESTED** [review]
One suggestion and tests for transactions would be nice, an example of sending a ping/pong tx to the node in the examples folder also will be helpful otherwise LGTM, once the change we discussed is done I'll approve

**Review by HDauven — DISMISSED** [review]

**Review by HDauven — APPROVED** [review]

File: execution-core/src/transfer/phoenix.rs
Comment (herr-seppia) [inline execution-core/src/transfer/phoenix.rs:279]: Why don't we flag the method as unsafe?

File: contracts/transfer/build.rs
Comment (herr-seppia) [inline contracts/transfer/build.rs:1]: Is this out of scope with the PR?

Please add some rationale or at least mention it in the PR description

File: rusk-prover/src/errors.rs
Comment (herr-seppia) [inline rusk-prover/src/errors.rs:55]: `{plonk_error:?}`

File: rusk-prover/src/lib.rs
Comment (herr-seppia) [inline rusk-prover/src/lib.rs:21]: This was under feature flag in order to let custom prover implementation to use prover lib without the need of local prover (check our lambda prover in `dusk_prover` repository)

File: rusk-prover/src/lib.rs
Comment (herr-seppia) [inline rusk-prover/src/lib.rs:28]: This is a breaking change I would avoid if we want to continue support our aws prover

**Review by herr-seppia — CHANGES_REQUESTED** [review]
Apart from some nit, I would like to keep the prover trait

File: execution-core/src/transfer/phoenix.rs
Comment (ZER0) [inline execution-core/src/transfer/phoenix.rs:279]: This is basically a setter, so idiomatically I would name this method `set_proof`.

File: execution-core/src/transfer/phoenix.rs
Comment (ZER0) [inline execution-core/src/transfer/phoenix.rs:264]: As discussed, it's better to have a recoverable error here instead of a panic.

File: rusk-prover/src/lib.rs
Comment (moCello) [reply rusk-prover/src/lib.rs:21]: Since the `rusk-prover::Prover` trait is superseded by `execution-core`s `Prove` trait, the library now only contains the `LocalProver` (which implements the `Prove` trait). That way the feature structure in rusk-prover is massively simplified.

File: rusk-prover/src/lib.rs
Comment (moCello) [reply rusk-prover/src/lib.rs:28]: Can't our aws prover use the `Prove` trait from execution-core?

**Review by ZER0 — APPROVED** [review]
A couple of things to change, but LGTM

File: wallet-core/src/transaction.rs
Comment (moCello) [reply wallet-core/src/transaction.rs:85]: type-alias for what?

Comment (Daksh14) [reply wallet-core/src/transaction.rs:85]:
```rust

type Input = (Note, Opening<_', _'>);`
type Inputs = Vec<Input>;

```

Comment (moCello) [reply wallet-core/src/transaction.rs:85]: Hm, I'm not sure if creating a type for that is worth it as sometimes a `Vec<Note, Opening<..>>` is needed but sometimes we also have a `Vec<Note, BlsScalar>` (for the nullifier) or even a `Vec<Note, Opening<..>, BlsScalar>`...

That said, I totally agree that this reads horribly.
I think the best thing to do is to re-export the `poseidon-merkle` types through `execution-core`. This which will already make this read much better (something like `Vec<Note, NoteOpening>`, `Vec<Note, NoteOpening, BlsScalar>`).

File: contracts/transfer/build.rs
Comment (moCello) [reply contracts/transfer/build.rs:1]: Done 👍

File: rusk-prover/src/lib.rs
Comment (herr-seppia) [reply rusk-prover/src/lib.rs:28]: I thought the trait was removed.
If it's just moved I'm ok with that

File: rusk/src/lib/http/prover.rs
Comment (herr-seppia) [inline rusk/src/lib/http/prover.rs:24]: I agree with this change, however this is a breaking change that would need a version update.

I suggest to keep the old topic to be backward compatible.

As soon as we move to RUES dispatch event we can choose the name we prefer, since it will be a breaking change anyway

**Review by herr-seppia — CHANGES_REQUESTED** [review]

File: execution-core/src/transfer/phoenix.rs
Comment (moCello) [reply execution-core/src/transfer/phoenix.rs]: It's not really unsafe in rust's sense. I added more substance to the comment though!

File: execution-core/src/transfer/phoenix.rs
Comment (moCello) [reply execution-core/src/transfer/phoenix.rs]: Should be solved now

**Review by herr-seppia — DISMISSED** [review]
LGTM

File: execution-core/src/error.rs
Comment (ZER0) [inline execution-core/src/error.rs:37]: I believe this should be `Execution` and not `Phoenix`.

**Review by ZER0 — DISMISSED** [review]
LGTM there is only a left over on the display of the error, I believe (check my comment)

**Review by ZER0 — APPROVED** [review]

**Comment by Daksh14** [comment]
LGTM

**Review by Daksh14 — APPROVED** [review]

**Pull request merged**
